### PR TITLE
fix(scripts/npc_calvin_montague) Added emotes when completing quests

### PR DESCRIPTION
**Pull request by fangshun2004**
## Changes Proposed:
1. Use the spell ID of sniffing
2. Factions that use sniffing change ids
3. Emotes are made when completing quest
4. Added Dead Player or Creature Away from Combat (Stay Attackable and Wait 3 Minutes to Reset)

[Video of the completion of the task on Blizzard's servers](https://clipchamp.com/watch/7Di3guvuoeJ)

This PR proposes changes to:
-  [ ] Core (units, players, creatures, game systems).
-  [x] Scripts (bosses, spell scripts, creature scripts).
-  [x] Database (SAI, creatures, etc).

## Issues Addressed:

- Closes 

## SOURCE:

The changes have been validated through:
- [x] Live research (checked on live servers, e.g Classic WotLK, Retail, etc.)
- [x] Sniffs (remember to share them with the open source community!)
- [x] Video evidence, knowledge databases or other public sources (e.g forums, Wowhead, etc.)
- [ ] The changes promoted by this pull request come partially or entirely from another project (cherry-pick). **Cherry-picks must be committed using the proper --author tag in order to be accepted, thus crediting the original authors, unless otherwise unable to be found**

## Tests Performed:

This PR has been:
- [x] Tested in-game by the author.
- [ ] Tested in-game by other community members/someone else other than the author/has been live on production servers.
- [ ] This pull request requires further testing and may have edge cases to be tested.

## How to Test the Changes:

- [x] This pull request can be tested by following the reproduction steps provided in the linked issue
- [ ] This pull request requires further testing. Provide steps to test your changes. If it requires any specific setup e.g multiple players please specify it as well.
1. Create a tribal role
2. .quest add 8
3. .go c 28474
5. Complete the quest
6. .go c 28493
7. Accept the quest
8. Reduces mob health to 30% and he should return to the home point and start making emotes Standing and using food

10. Delete the quest and start it again
11. After the battle starts, using .gm on, the player dies, leaves the battle, etc. When the mob returns, it should be attackable, and when the mob reaches the home point, it should be passive and can attack, and after waiting for a few minutes, the mob name turns green and the non-attackable state

## Known Issues and TODO List:

- [x]After defeating a mob, the mob uses spell 7737 with no visual effect (because SetStandState 0 and cast are in the same timestamp in the sniffing data, and the spell stops immediately after standing. I don't know how Blizzard did it)
- [ ]

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: src/server/scripts/EasternKingdoms/zone_tirisfal_glades.cpp [hunk 1]
```diff
@@ -18,15 +18,13 @@
...
+SDComment: Quest support: 590
 SDCategory: Tirisfal Glades
 EndScriptData */
 
 /* ContentData
 npc_calvin_montague
-go_mausoleum_door
-go_mausoleum_trigger
-EndContentData */
+*/
```
Comment (Kitzunu) [inline src/server/scripts/EasternKingdoms/zone_tirisfal_glades.cpp:27]: Just delete this

File: src/server/scripts/EasternKingdoms/zone_tirisfal_glades.cpp [hunk 2]
```diff
@@ -38,134 +36,161 @@ EndContentData */
...
+    EVENT_TALK = 2,
+    EVENT_DRINK_AND_QUEST_COMQUEST = 3,
+    EVENT_END = 4,
 };
 
 class npc_calvin_montague : public CreatureScript
 {
 public:
-    npc_calvin_montague() : CreatureScript("npc_calvin_montague") { }
+    npc_calvin_montague() : CreatureScript("npc_calvin_montague") {}
```
Comment (Kitzunu) [inline src/server/scripts/EasternKingdoms/zone_tirisfal_glades.cpp:55]: if you are rewriting a script this much, just use the new registry

File: src/server/scripts/EasternKingdoms/zone_tirisfal_glades.cpp [hunk 3]
```diff
@@ -38,134 +36,161 @@ EndContentData */
...
     {
         if (quest->GetQuestId() == QUEST_590)
         {
-            creature->SetFaction(FACTION_ENEMY);
-            creature->SetImmuneToPC(false);
-            CAST_AI(npc_calvin_montague::npc_calvin_montagueAI, creature->AI())->AttackStart(player);
+            creature->SetFaction(FACTION_MONSTER);
+            creature->RemoveNpcFlag(UNIT_NPC_FLAG_QUESTGIVER);
+            creature->RemoveUnitFlag(UNIT_FLAG_IMMUNE_TO_PC);//This is reset when out of combat
+            creature->SetUnitFlag(UNIT_FLAG_PET_IN_COMBAT);
```
Comment (Kitzunu) [inline src/server/scripts/EasternKingdoms/zone_tirisfal_glades.cpp:64]: Is this sniffed? this looks weird to say the least

File: src/server/scripts/EasternKingdoms/zone_tirisfal_glades.cpp [hunk 4]
```diff
@@ -38,134 +36,161 @@ EndContentData */
...
-            if (me->GetVictim() || me->IsFriendlyTo(pAttacker))
-                return;
+            if (m_uiPhase == 1)
+            {
+                events.ScheduleEvent(EVENT_EMOTE, 2500ms);
+            }
+            else
+            {
+                events.ScheduleEvent(EVENT_END, 3min);
+            }
```
Comment (Kitzunu) [inline src/server/scripts/EasternKingdoms/zone_tirisfal_glades.cpp:93]: just remove the brackets

File: src/server/scripts/EasternKingdoms/zone_tirisfal_glades.cpp [hunk 5]
```diff
@@ -38,134 +36,161 @@ EndContentData */
...
         {
-            m_uiPhase = 0;
-            m_uiPhaseTimer = 5000;
-            m_uiPlayerGUID.Clear();
-
-            me->RestoreFaction();
-
-            if (!me->IsImmuneToPC())
-                me->SetImmuneToPC(true);
+            if (m_uiPhase != 1)
```
Comment (Kitzunu) [inline src/server/scripts/EasternKingdoms/zone_tirisfal_glades.cpp:77]: m_uiPhase can be either 0 or 1, so use 
if (m_uiPhase) - This means true if m_uiPhase exist, aka any positive number
if (!m_uiPhase) - This means true if m_uiPhase does not exist, aka 0

File: src/server/scripts/EasternKingdoms/zone_tirisfal_glades.cpp [hunk 6]
```diff
@@ -38,134 +36,161 @@ EndContentData */
...
+                        me->SetStandState(UNIT_STAND_STATE_STAND);
                         if (Player* player = ObjectAccessor::GetPlayer(*me, m_uiPlayerGUID))
                             player->AreaExploredOrEventHappens(QUEST_590);
-
-                        DoCast(me, SPELL_DRINK, true);
-                        ++m_uiPhase;
+                        events.ScheduleEvent(EVENT_END, 12900ms);
+                        break;
+                    case EVENT_END:
+                        //me->SetStandState(UNIT_STAND_STATE_STAND);
```
Comment (Kitzunu) [inline src/server/scripts/EasternKingdoms/zone_tirisfal_glades.cpp:166]: if we dont use it we delete it.

File: src/server/scripts/EasternKingdoms/zone_tirisfal_glades.cpp (hunk 3 above)
Comment (fangshun2004) [reply src/server/scripts/EasternKingdoms/zone_tirisfal_glades.cpp:64]:
```
ServerToClient: SMSG_UPDATE_OBJECT (0x27CB) Length: 169 ConnIdx: 1 Time: 12/03/2023 09:15:18.199 Number: 74959
NumObjUpdates: 2
MapID: 0 (0)
HasRemovedObjects: False
Data size: 158
[0] UpdateType: Values
[0] Object Guid: Full: 0x000000000000000000000001 Player/0 R4904/S0 Map: 0 Low: 13355676
[0] Health: 13872
[0] Flags: 526344
[0] PvpFlags: 1
[0] PlayerFlags: 256
[1] UpdateType: Values
[1] Object Guid: Full: 0x000000000000000000000001 Creature/0 R4908/S154 Map: 0 Entry: 6784 Low: 6970167
[1] EntryID: 6784
[1] DynamicFlags: 0
[1] Health: 102
[1] MaxHealth: 102
[1] DisplayID: 5509
[1] StateSpellVisualID: 0
[1] StateAnimID: 1772
[1] StateAnimKitID: 0
[1] Target: Full: 0x000000000000000000000001 Player/0 R4904/S0 Map: 0 Low: 13355676
[1] FactionTemplate: 14
[1] Flags: 559616
[1] [0] NpcFlags: 0
[1] [1] NpcFlags: 0
```
[0] Flags: 526344 and [1] Flags: 559616  Is it?

File: src/server/scripts/EasternKingdoms/zone_tirisfal_glades.cpp (hunk 4 above)
Comment (fangshun2004) [reply src/server/scripts/EasternKingdoms/zone_tirisfal_glades.cpp:93]: I'm sorry, I don't quite understand what this means
 `if (m_uiPhase == 1)`
Modify to
 `if (m_uiPhase)`
？？？

File: src/server/scripts/EasternKingdoms/zone_tirisfal_glades.cpp (hunk 3 above)
Comment (Kitzunu) [reply src/server/scripts/EasternKingdoms/zone_tirisfal_glades.cpp:53]: 2048 is UNIT_FLAG_PET_IN_COMBAT

File: src/server/scripts/EasternKingdoms/zone_tirisfal_glades.cpp [hunk 7]
```diff
@@ -38,134 +25,160 @@ EndContentData */
...
+                        m_uiPlayerGUID = owner->GetGUID();
                     }
                 }
             }
         }
 
         void UpdateAI(uint32 diff) override
         {
-            if (m_uiPhase)
+            if (!m_uiPhase&& me->HealthBelowPct(30))
```
Comment (Kitzunu) [inline src/server/scripts/EasternKingdoms/zone_tirisfal_glades.cpp:121]:
```suggestion
            if (!m_uiPhase && me->HealthBelowPct(30))
```

File: src/server/scripts/EasternKingdoms/zone_tirisfal_glades.cpp (hunk 4 above)
Comment (Kitzunu) [reply src/server/scripts/EasternKingdoms/zone_tirisfal_glades.cpp]:
```cpp
            if (m_uiPhase)
                events.ScheduleEvent(EVENT_EMOTE, 2500ms);
            else
                events.ScheduleEvent(EVENT_END, 3min);

File: src/server/scripts/EasternKingdoms/zone_tirisfal_glades.cpp (hunk 3 above)
Comment (fangshun2004) [reply src/server/scripts/EasternKingdoms/zone_tirisfal_glades.cpp:53]: image

File: src/server/scripts/EasternKingdoms/zone_tirisfal_glades.cpp (hunk 2 above)
Comment (Kitzunu) [reply src/server/scripts/EasternKingdoms/zone_tirisfal_glades.cpp:44]: This still valid. Use RegisterCreatureAI instead

Comment (fangshun2004) [reply src/server/scripts/EasternKingdoms/zone_tirisfal_glades.cpp:44]: I've looked at other scripts and tried to use
 ```
void AddSC_tirisfal_glades()
{   
RegisterCreatureAI(npc_calvin_montague);
}
```
But
` bool OnQuestAccept(Player* player, Creature* creature, Quest const* quest) override `
is not used in other scripts
OnQuestAccept cannot be used in other classes

Comment (fangshun2004) [reply src/server/scripts/EasternKingdoms/zone_tirisfal_glades.cpp:44]: I've looked at other scripts and tried to use
 ```
void AddSC_tirisfal_glades()
{   
RegisterCreatureAI(npc_calvin_montague);
}
```
But
` bool OnQuestAccept(Player* player, Creature* creature, Quest const* quest) override `
is not used in other scripts
OnQuestAccept cannot be used in other classes

**Comment by sudlud** [comment]
what's the state of this PR? is this rady to be tested? 🙂

can you please also mark the open reviews as resolved if you've already updated the PR accordingly

**Comment by fangshun2004** [comment]
Admin requires RegisterCreatureAI to register scripts . As commented above, I can't use RegisterCreatureAI to register scripts, so I don't know how to modify this script next.

**Comment by fangshun2004** [comment]
I'm having trouble converting this script to smart script.
image
On Reset (Phase1) On Home Reached (Phase 1) On Home Reached (Phase 2) event could not be triggered

**Comment by fangshun2004** [comment]
This one has been done

**Comment by sudlud** [comment]
> This one has been done

Thanks for the update!

How does this PR relate to #18096? Would merging this make #18096 obsolete or are they covering different topics actually?

File: src/server/scripts/EasternKingdoms/zone_tirisfal_glades.cpp
Comment (sudlud) [inline src/server/scripts/EasternKingdoms/zone_tirisfal_glades.cpp:1]: You can most likely delete this file completely

**Comment by fangshun2004** [comment]
> > This one has been done
> 
> Thanks for the update!
> 
> How does this PR relate to #18096? Would merging this make #18096 obsolete or are they covering different topics actually?

#18096 is to move the original C++ script to SmartScript. This RP is made with sniffing data and video (the video link is out of date) with some details added (emotes and spell ids and disengagement), which will make #18096 obsolete after the update

File: src/server/scripts/EasternKingdoms/zone_tirisfal_glades.cpp
Comment (fangshun2004) [reply src/server/scripts/EasternKingdoms/zone_tirisfal_glades.cpp:1]: In the Tirisfar Woodlands, it is also necessary to implement the script for the current location query of the transport ship (gossip and boarding broadcast by the transport ship manager)

**Comment by elthehablo** [comment]
Oh nice work on doing this because I never got around to fixing mine. Yes you can delete the zone script as well as its initialization

**Comment by sudlud** [comment]
Please see my review above - the lines that you commented can be removed

**Review by elthehablo — APPROVED** [review]

File: data/sql/updates/pending_db_world/rev_1704025138911597700.sql [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+(@ENTRY * 100 + 1, 9, 12, 0, 0, 0, 100, 0, 0, 0, 0, 0, 42, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 'Calvin Montague - action list - Set invincibility to 0 HP'),
+(@ENTRY * 100 + 1, 9, 13, 0, 0, 0, 100, 0, 0, 0, 0, 0, 211, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 'Calvin Montague - action list - Allow SAI phase reset'),
+(@ENTRY * 100 + 1, 9, 14, 0, 0, 0, 100, 0, 0, 0, 0, 0, 22, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 'Calvin Montague - action list - Set event phase to default (0)');
+
+DELETE
+FROM `quest_details`
+WHERE `ID` IN (590,8);
+INSERT INTO `quest_details` (`ID`, `Emote1`, `Emote2`, `Emote3`, `Emote4`, `EmoteDelay1`, `EmoteDelay2`, `EmoteDelay3`, `EmoteDelay4`, `VerifiedBuild`) VALUES
+(8, 2, 1, 1, 0, 0, 0, 0, 0, 52237),
+(590, 6, 1, 14, 0, 0, 0, 0, 0, 52237);
```
Comment (sudlud) [inline data/sql/updates/pending_db_world/rev_1704025138911597700.sql:56]: missing new line

**Pull request merged**